tests/integration/pilot/common/traffic.go

```diff
@@ -40,18 +42,73 @@ type TrafficCall struct {
 }
 
 type TrafficTestCase struct {
-	name   string
+	name string
+	// config can optionally be templated using the params src, dst (each are []echo.Instance)
 	config string
 
 	// Multiple calls. Cannot be used with call/opts
 	children []TrafficCall
 
 	// Single call. Cannot be used with children.
 	call func(t test.Failer, options echo.CallOptions, retryOptions ...retry.Option) echoclient.ParsedResponses
+	// opts specifies the echo call options. When using RunForApps, the Target will be set dynamically.
 	opts echo.CallOptions
+	// validate is used to build validators dynamically when using RunForApps based on the active/src dest pair
+	validate func(src echo.Instance, dst echo.Instances) echo.Validator
 
 	// setting cases to skipped is better than not adding them - gives visibility to what needs to be fixed
 	skip bool
+
+	// workloadAgnostic is a temporary setting to trigger using RunForApps
+	// TODO remove this and force everything to be workoad agnostic
+	workloadAgnostic bool
+}
+
+func (c TrafficTestCase) RunForApps(t framework.TestContext, apps echo.Instances, namespace string) {
+	if c.opts.Target != nil {
+		t.Fatal("TrafficTestCase.RunForApps: opts.Target must not be specified")
+	}
+	if c.call != nil {
+		t.Fatal("TrafficTestCase.RunForApps: call must not be specified")
+	}
+	// just check if any of the required fields are set
+	optsSpecified := c.opts.Port != nil || c.opts.PortName != "" || c.opts.Scheme != ""
+	if optsSpecified && len(c.children) > 0 {
+		t.Fatal("TrafficTestCase: must not specify both opts and children")
+	}
+
+	echotest.New(t, apps).
+		SetupForPair(func(t framework.TestContext, src, dst echo.Instances) error {
+			cfg := tmpl.MustEvaluate(
+				yml.MustApplyNamespace(t, c.config, namespace),
+				map[string]echo.Instances{
+					"src": src,
+					"dst": dst,
+				},
+			)
+			return t.Config().ApplyYAML("", cfg)
+		}).
+		From(echotest.OnRegularPodSource).
+		To(echotest.OneRegularPodDestination, echotest.ReachableDestinations).
```

Comment:
Should ReachableDestinations be before OneRegularPodDestination? Otherwise, couldn't OneRegularPodDestination pick a single one, which isn't reachable?

In practice probably not an issue though

Reply:
good point, fixed

Reply:
we should probably wrap the common combinations of filters together